### PR TITLE
157 - Added more prometheus metrics

### LINKED ISSUE #157: RIG specific prometheus metrics
This issue is supposted to cover implementation of RIG specifc metrics.

1. implement "setup" of metrics within the instrumenter modules under `rig_metrics/lib/rig_metrics/*` Those are already implemented for the planned metrics, but commented out right now to avoid exposing not working metrics
2. implement metrics triggers for the varying use cases in the other otp apps (might differ per metrics). For this a dependency to rig_metrics will be required in the other otp apps

**Metrics planned as of latest discussion:**

RigProxyInstrumenter

* rig_current_session_count --> As Prometheus Gauge
* rig_current_subscription_count --> As Prometheus Gauge
* rig_current_open_proxy_connection_count --> As Prometheus Gauge
* rig_proxy_requests_total --> As Prometheus Counter

RigEventhubInstrumenter

* rig_processed_events_total --> As Prometheus counter + Label Node (e.g. Kafka, etc)
* rig_forwarded_events_total --> As Prometheus counter + Label Node (e.g. Kafka, etc)
* rig_dropped_events_total --> As Prometheus counter + Label Node (e.g. Kafka, etc)

RigControlInstrumenter

* rig_current_session_blacklisted_count --> As Prometheus Gauge
* rig_current_routes_configured_count --> As Prometheus Gauge

**Further Metric ideas:**
- Messages_total
- Messages_Error
- Latency
- Throughput
- etc.

### PR DESCRIPTION
**Pull request by mmacai**
## Description

- added Prometheus metrics for events, subscriptions and blacklist
- added custom Grafana dashboard
- added local docker-compose setup for Prometheus and Grafana

Closes #157 #222 

## What to look out for

Dear reviewer, I want you to

- gain knowledge - I think it's important for you to know about the change.
- check that the code works on your machine.
- suggest implementation-code design/structure/readability improvements.
- suggest test-code design/structure/readability improvements.
- suggest documentation improvements.
- suggest language/writing improvements.
- help to find bugs (described below!).

Notes:

- go through metrics docs -> should be enough to understand what it can do and how to use it

@jakobwgnr you can check it as well.

## Process

The goal is to improve not only the code in this PR but also our skills! The "rules":

- The review is considered "done" as soon as all reviewers have added their review, and all their comments have been addressed.
- For knowledge-sharing reviews, each reviewer should "approve" the PR after studying its content.
- After the approval, the merge is concluded by the developer.

Have fun!

### CODE REVIEW

File: docs/metrics-details.md [hunk 1]
```diff
@@ -36,38 +39,136 @@ Following Labels are provided:
...
 We are exposing the standard metrics providing information about the running application itself.
 
 Following standard metrics are currently provided:
-- erlang_vm_memory_ets_tables - *Erlang VM ETS Tables count*
-- erlang_vm_memory_atom_bytes_total (Labels: Free, Used) - *The total amount of memory currently allocated for atoms. This memory is part of the memory presented as system memory.*
-- erlang_vm_port_limit - *The maximum number of simultaneously existing ports at the local node*
+
+- erlang*vm_memory_ets_tables - \_Erlang VM ETS Tables count*
+- erlang*vm_memory_atom_bytes_total (Labels: Free, Used) - \_The total amount of memory currently allocated for atoms. This memory is part of the memory presented as system memory.*
+- erlang*vm_port_limit - \_The maximum number of simultaneously existing ports at the local node*
```
Comment (kevinbader) [inline docs/metrics-details.md:146]: formatting seems off - `\_ ... *` ?

File: lib/rig/event_filter.ex [hunk 2]
```diff
@@ -284,21 +285,23 @@ defmodule Rig.EventFilter do
...
   end
 
   # ---
 
-  @callback forward_event(CloudEvent.t()) :: :ok
-  @spec forward_event(CloudEvent.t()) :: :ok
-  def forward_event(%CloudEvent{} = event) do
+  @callback forward_event(CloudEvent.t(), String.t(), String.t()) :: :ok
+  @spec forward_event(CloudEvent.t(), String.t(), String.t()) :: :ok
+  def forward_event(%CloudEvent{} = event, source, topic) do
```
Comment (kevinbader) [inline lib/rig/event_filter.ex:299]: Obviously you'd like to know `source` and `topic` for the metrics, but I'm worried about what this means in terms of semantics. The meaning of `forward_event` is (has been?): "take an event, forward it to any subscribers". To a RIG subscriber, `topic` doesn't mean anything, and also subscribing to an event does not depend on the source. I'd prefer `forward_event` to not take source and topic and arguments. Instead, I'd rather not include that in the metric. If the goal is counting events per Kafka topic, it should be counted in the Kafka handler.

Comment (mmacai) [reply lib/rig/event_filter.ex:299]: I'll take a look, my idea was to be able to monitor something like this (events):

Consumer:

- number of events consumed in total
- number of events that were skipped/ignored for some reason
- number of events that failed to be consumed
- number of events successfully forwarded to a subscription via whatever channel (ws, sse, lp)

Producer:

- number of successfully produced events
- number of events that failed to be produced

As a user when I want to create some dashboards I need some tags to be able to filter by some criteria:

- topic/namespace whatever we call it
- source - kafka, kinesis, etc.

These were my thoughts back then, but tbh have to go through it one more time and check what make sense as we changed a lot of thing in the meantime.

**Comment by mmacai** [comment]
@kevinbader Overhauled some parts, you can give it another shot.

**Comment by mmacai** [comment]
@kevinbader Did some adjustments and merged with master, should be ready for another review.

**Review by kevinbader — APPROVED** [review]
🎉

**Pull request merged**